README
PADRES v2.0
===========

PADRES is a distributed content-based pub/sub middleware. This
middleware can be used to create an acyclic broker overlay to route
content-based pub/sub between publishers and subscribers. The
following components included in this package are:

1. broker: to create the distributed routing backbone.
2. client: to connect to a broker and acts as a publisher or subscriber. 
3. monitor: to monitor the deployed PADRES system.


Files and Directories:
---------------------
1. README: this file.
2. INSTALL: installation instructions.
2. ISSUES: open issues.
3. COPYRIGHT: copyright notice.
4. install.sh: script to install PADRES in the system.
5. bin/: scripts to run PADRES components. 
6. build/: the PADRES component binaries. 
7. etc/: PADRES configuration files.
8. demo/: various demo scripts instructions.
9. doc/: Generated Javadocs.


Installation:
-------------
Check the instructions in the INSTALL file.


Documentation
-------------
The PADRES user guide is available at
http://www.msrg.org/projects/padres/docs/user_guide

The PADRES developer guide is available at 
http://www.msrg.org/projects/padres/docs/dev_guide


--------------------------------------------------------------------
Middleware Systems Research Group
University of Toronto
2012-02-11


CHANGELOG
PADRES v2.0
===========

This version of PADRES includes a number of new features, architectural changes,
demos, and performance enhancements as outlined below.

New features
- Support for socket communication between brokers and clients.
- Brokers support advertisement covering.
- Brokers can be controlled with command line and Web management interfaces.
- Historic database can use an embedded Apache Derby database in addition to 
  an external user-provisioned database.
- Monitor displays cleaner broker and client names.
- Monitor supports tracing of messages.
- Composite subscription can support subexpressions in both the left and right 

Architectural changes
- Communication layer can support pluggable protocols.
- Broker and client URI formats are generalized for pluggable protocols.
- More robust parsing of message formats based on a JavaCC parser.
- More flexible configuration settings.
- Client is easier to extend and process publications.
- Client now supports a pluggable command handler.

Demos
- Stock quote
- Web client
- Historic
- Workflow
- Failure detector and monitor
- Fixed cycles
- Dynamic cycles

Performance enhancements
- Faster and more robust test framework based on dependency injection pattern.
- Faster subscription covering algorithm using a more light-weight covering 
  data structure.
- Faster publication forwarding by pruning search space.


--------------------------------------------------------------------
Middleware Systems Research Group
University of Toronto
2012-02-11


install_binary.sh
#!/bin/sh

USAGE="Usage: $(basename $0) [-h] [-d padres_dir]"

INSTALL_DIR="/usr/local/"
DOWNLOAD_DIR="$(cd $(dirname $0) && pwd)"

# process the command line options
while [ $# -gt 0 ]; do
	case "$1" in
		( -h )
			echo $USAGE
			exit 0 ;;
		( -d )
			INSTALL_DIR="$2"
			shift 2 ;;
		( * )
			echo "invalid option $1"
			echo $USAGE
			exit 1 ;;
    esac
done

# remove the '/' at the end, if any
INSTALL_DIR=`echo $INSTALL_DIR | sed s:/$::`

# create a 'padres' directory inside the given directory
INSTALL_DIR="${INSTALL_DIR}/padres"
# create the installation directory if it does not exist
# exit, if permission denied
if [ ! -d $INSTALL_DIR ]; then
	echo "creating the instation directory <$INSTALL_DIR>..."
	mkdir $INSTALL_DIR
	# check the exit status of the mkdir; exit, if mkdir is failed
	if [ $? != "0" ]; then
	    exit 1
	fi
fi

# install binaries, scripts, and config files
# basically copy them to the INSTALL_DIR
echo "copying files to the installation directory..."
cp -r $DOWNLOAD_DIR/build/ $INSTALL_DIR
cp -r $DOWNLOAD_DIR/bin/ $INSTALL_DIR
cp -r $DOWNLOAD_DIR/demo/ $INSTALL_DIR
cp -r $DOWNLOAD_DIR/etc/ $INSTALL_DIR

#give execution permissions to the scripts
chmod 755 $INSTALL_DIR/bin/*.sh

echo "PADRES installation complete"
echo "\nPlease set/update environment variables using following commands:"
echo ">>export PADRES_HOME=\"$INSTALL_DIR\""
echo ">>export PATH=\"\$PATH:\$PADRES_HOME/bin\""
exit 0



demo/src/ca/utoronto/msrg/padres/demo/stockquote/StockSubscriber.java
package ca.utoronto.msrg.padres.demo.stockquote;

import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.rmi.RemoteException;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.TimeZone;

import javax.swing.Timer;

import org.apache.log4j.Logger;

import ca.utoronto.msrg.padres.client.Client;
import ca.utoronto.msrg.padres.client.ClientConfig;
import ca.utoronto.msrg.padres.client.ClientException;
import ca.utoronto.msrg.padres.common.message.Message;
import ca.utoronto.msrg.padres.common.message.Publication;
import ca.utoronto.msrg.padres.common.message.PublicationMessage;
import ca.utoronto.msrg.padres.common.message.Subscription;
import ca.utoronto.msrg.padres.common.message.SubscriptionMessage;
import ca.utoronto.msrg.padres.common.message.parser.MessageFactory;
import ca.utoronto.msrg.padres.common.message.parser.ParseException;
import ca.utoronto.msrg.padres.common.util.CommandLine;
import ca.utoronto.msrg.padres.common.util.LogException;
import ca.utoronto.msrg.padres.common.util.LogSetup;

// import loadbalancing.mediator.Mediator;

/**
 * A subscriber that subscribes to stock quotes and logs delivery delay to file. Logging is synced
 * at the beginning of each second (almost in fact).
 * 
 * @author cheung
 * 
 */
public class StockSubscriber extends StockClient {

	// Constants
	private final static short MAX_DUPLICATE_HISTORY_SIZE = 16;

	private final static String DEFAULT_SUBSCRIBER_ID = "Sub-" + DEFAULT_CLIENT_ID;

	private final static long DELAY_BETWEEN_NEW_SUBSCRIPTION = 1000; // ms

	// Command-line argument-related constants
	private static final String CMD_ARG_SUBSCRIPTION = "s";

	public static final String SUB_DELIMIT_CHAR = "/";

	private static final int SUB_FACTOR = 10;

	// For duplicate detection Accounting
	private final LinkedHashMap<Publication, Object> messageHistory;

	// Synchronization/mutex/locks
	private final Object avgDelaySync = new Object();

	private final Object logSync = new Object();

	// Holds this subscriber's subscription
	private Map<String, Subscription> subId2SubMap;

	// Logging
	private final Timer loggingTimer;

	private int duplicates;

	// Performance Measurements
	private double minDelay;

	private double maxDelay;

	private long received;

	private double cumulativeDelay;
	
	private String brokerId;

	/**
	 * Constructor
	 * 
	 * @param id
	 * @param strSubscription
	 * @throws ClientException
	 * @throws RemoteException
	 */
	public StockSubscriber(ClientConfig clientConfig, String strSubscription)
			throws ClientException {
		super(clientConfig);
		brokerId = clientConfig.connectBrokerList[0];
		// send subscriptions
		sendSubscriptions(strSubscription);
		// duplicate detection stuff
		duplicates = 0;
		messageHistory = new LinkedHashMap<Publication, Object>(MAX_DUPLICATE_HISTORY_SIZE,
				(float) 0.75, true) {

			private static final long serialVersionUID = 1L;

			// We want to limit the history size to MAX_HISTORY_SIZE
			protected boolean removeEldestEntry(Map.Entry<Publication, Object> eldest) {
				return size() > MAX_DUPLICATE_HISTORY_SIZE;
			}
		};
		// performance metrics
		minDelay = Double.MAX_VALUE;
		maxDelay = Double.MIN_VALUE;
		received = 0;
		cumulativeDelay = 0;
		// For logging
		loggingTimer = setupLoggingTimer();
	}

	public double getMinDelay() {
		return minDelay;
	}

	public void setMinDelay(double value) {
		minDelay = value;
	}

	public double getMaxDelay() {
		return maxDelay;
	}

	public void setMaxDelay(double value) {
		maxDelay = value;
	}

	private boolean sendSubscriptions(String strSubscription) throws ClientException {
		// Parse the subscription string
		String[] subStrList = strSubscription.split(SUB_DELIMIT_CHAR);
		List<Subscription> subList = new ArrayList<Subscription>();
		for (String subStr : subStrList) {
			Subscription sub;
			try {
				sub = MessageFactory.createSubscriptionFromString(subStr);
			} catch (ParseException e) {
				exceptionLogger.error(e.getMessage());
				throw new ClientException(e);
			}
			subList.add(sub);
		}
		// subscribe to all the brokers with all the given subscriptions
		subId2SubMap = new HashMap<String, Subscription>();
		for (String brokerURI : getBrokerURIList())
			for (Subscription sub : subList) {
				for(int i = 0; i < SUB_FACTOR; i++){
					SubscriptionMessage subMsg = subscribe(sub, brokerURI);
					subId2SubMap.put(subMsg.getMessageID(), sub);
					try {
						Thread.sleep(DELAY_BETWEEN_NEW_SUBSCRIPTION);
					} catch (Exception e) {
					}
				}
			}
		return true;
	}

	/*
	 * Merely just instantiating the timer object, have not started the timer yet!
	 */
	private Timer setupLoggingTimer() {
		ActionListener taskPerformer = new ActionListener() {

			public void actionPerformed(ActionEvent evt) {
				synchronized (logSync) {
					logSync.notify();
				}
			}
		};
		return new Timer(clientConfig.logPeriod * 1000, taskPerformer);
	}

	/*
	 * Called by the timer to do logging whenever the logging period is up.
	 */
	private void doLogging() {
		// Calculate the average delay
		double avgDelay = 0;

		synchronized (avgDelaySync) {
			if (received > 0) {
				avgDelay = cumulativeDelay / (double) received;
				cumulativeDelay = 0;
				received = 0;
				// Can't compute the average if no messages were received
			} else {
				avgDelay = 0;
			}
			avgDelaySync.notifyAll();
		}

		clientLogger.info(Math.round(Calendar.getInstance(TimeZone.getTimeZone("GMT-5")).getTimeInMillis() / 1000.0)
				+ "\t"
				+ avgDelay
				+ "\t"
				+ (minDelay == Double.MAX_VALUE ? 0 : minDelay)
				+ "\t"
				+ (maxDelay == Double.MIN_VALUE ? 0 : maxDelay) + "\t" + duplicates);

		// reset some metrics
		minDelay = Double.MAX_VALUE;
		maxDelay = Double.MIN_VALUE;
	}

	/*
	 * (non-Javadoc)
	 * 
	 * @see stockquote.StockClient#handlePub(message.Publication)
	 */
	public void handlePub(Publication pub) {
		String topic = pub.getClassVal();
		// Only log stock quote publication messages
		if (topic.equalsIgnoreCase("STOCK")) {
			handleStockQuotePub(pub);
			// } else if (topic.equalsIgnoreCase(Mediator.CLASS_CLIENT_CONTROL)) {
			// handleSubscriberControlPub(pub);
		} else {
			clientLogger.warn("Got unexpected publication message: " + pub);
		}
	}

	/**
	 * Nothing else except do periodic logging
	 * 
	 */
	public void run() {
		// write column headers into log file
		// logger.info("# Avg delay | Min delay | Max delay | Duplicates");
		syncStartLoggingTimer();

		// Wait till the period has come to log
		synchronized (logSync) {
			while (true) {
				try {
					logSync.wait();
				} catch (InterruptedException e) {
				}
				doLogging();
			}
		}
	}

	public void processMessage(Message msg) {
		if (msg instanceof PublicationMessage) {
			if(Message.getBrokerIdFromMsgId(msg.getMessageID()).equals(brokerId)){
				Publication pub = ((PublicationMessage) msg).getPublication();
				long currTime = System.currentTimeMillis();
				long timestamp = pub.getTimeStamp().getTime();
				long delay = currTime - timestamp;
				System.out.println(pub.getPubID() + "\t" + delay);
			} else {
				Publication pub = ((PublicationMessage) msg).getPublication();
				System.out.println(pub.getPubID() + "\t" + "-1");
			}
		}
	}

	/*
	 * Want to sync the logging to the start of every minute
	 */
	private void syncStartLoggingTimer() {
		Calendar now = Calendar.getInstance(TimeZone.getTimeZone("GMT-5"));
		int seconds = now.get(Calendar.SECOND);
		int milliseconds = now.get(Calendar.MILLISECOND);
		long waitTimeInMs = ((59 - seconds) * 1000) + (1000 - milliseconds);
		try {
			Thread.sleep(waitTimeInMs);
		} catch (Exception e) {
			e.printStackTrace();
		}
		loggingTimer.start();
	}

	/*
	 * (non-Javadoc)
	 * 
	 * @see universal.rmi.RMIUniversalClient#shutdown()
	 */
	public void shutdown() {
		loggingTimer.stop();
		// doLogging();
	}

	private static void showUsage() {
		System.out.printf("Usage: java StockPublisher -%s <id> -%s <broker_uri>\n\t"
				+ "-%s <subscription>\n", ClientConfig.CLI_OPTION_ID,
				ClientConfig.CLI_OPTION_BROKER_LIST, CMD_ARG_SUBSCRIPTION);
	}

	private void handleStockQuotePub(Publication pub) {
		// Update performance results
		// convert from ms to s
		double delay = (Calendar.getInstance(TimeZone.getTimeZone("GMT-5")).getTimeInMillis() - pub.getTimeStamp().getTime()) / 1000.0;
		minDelay = Math.min(minDelay, delay);
		maxDelay = Math.max(maxDelay, delay);

		// Update variables for computing the average
		synchronized (avgDelaySync) {
			cumulativeDelay += delay;
			received++;
			avgDelaySync.notifyAll();
		}

		clientLogger.debug("Got publication " + pub + " with delay " + delay);

		System.out.println(pub.getTimeStamp().getTime() + "\t" + delay);

		// Detect for duplcates
		if (messageHistory.containsKey(pub)) {
			clientLogger.debug("Publication is duplicate!");
			duplicates++;
			messageHistory.remove(pub);
		} else {
			messageHistory.put(pub, null);
		}
	}

	/**
	 * @param args
	 */
	public static void main(String[] args) {
		String[] cliKeys = ClientConfig.getCommandLineKeys();
		String[] fullCLIKeys = new String[cliKeys.length + 1];
		System.arraycopy(cliKeys, 0, fullCLIKeys, 0, cliKeys.length);
		fullCLIKeys[cliKeys.length] = CMD_ARG_SUBSCRIPTION + ":";
		CommandLine cmdLine = new CommandLine(fullCLIKeys);
		try {
			cmdLine.processCommandLine(args);
		} catch (Exception e) {
			e.printStackTrace();
			showUsage();
			System.exit(1);
		}
		String subscription = cmdLine.getOptionValue(CMD_ARG_SUBSCRIPTION);

		try {
			new LogSetup(null);
		} catch (LogException e1) {
			e1.printStackTrace();
			System.exit(1);
		}

		Logger logger = Logger.getLogger(Client.class);
		try {
			ClientConfig clientConfig = new ClientConfig();
			clientConfig.clientID = DEFAULT_SUBSCRIBER_ID;
			clientConfig.overwriteWithCmdLineArgs(cmdLine);
			if (subscription == null) {
				logger.fatal("Don't know what to subscribe because subscription is not specified.");
				showUsage();
				System.exit(1);
			} else if (clientConfig.connectBrokerList == null
					|| clientConfig.connectBrokerList.length == 0) {
				logger.fatal("Don't know which broker to connect to.");
				showUsage();
				System.exit(1);
			}
			StockSubscriber subscriber = new StockSubscriber(clientConfig, subscription);
			logger.info("Stock subscriber created. Subscribing to " + subscription);
			// Let it run forever until this program is killed
			subscriber.run();
		} catch (ClientException e) {
			e.printStackTrace();
			System.exit(1);
		}
	}

}


demo/src/ca/utoronto/msrg/padres/demo/stockquote/StockPublisher.java
package ca.utoronto.msrg.padres.demo.stockquote;

import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.HashMap;

import org.apache.log4j.Logger;

import ca.utoronto.msrg.padres.client.Client;
import ca.utoronto.msrg.padres.client.ClientConfig;
import ca.utoronto.msrg.padres.client.ClientException;
import ca.utoronto.msrg.padres.common.message.Advertisement;
import ca.utoronto.msrg.padres.common.message.AdvertisementMessage;
import ca.utoronto.msrg.padres.common.message.Predicate;
import ca.utoronto.msrg.padres.common.message.Publication;
import ca.utoronto.msrg.padres.common.message.PublicationMessage;
import ca.utoronto.msrg.padres.common.message.parser.MessageFactory;
import ca.utoronto.msrg.padres.common.message.parser.ParseException;
import ca.utoronto.msrg.padres.common.util.CommandLine;
import ca.utoronto.msrg.padres.common.util.LogException;
import ca.utoronto.msrg.padres.common.util.LogSetup;

/**
 * 
 * @author cheung
 * 
 *         Publishes stock quote publications just like the publisher client in the simulator
 * 
 *         Features: - delivery delay logging - does not shutdown if publisher is not connected with
 *         anyone
 * 
 */
public class StockPublisher extends StockClient implements Runnable {

	// CONSTANTS
	private static final String DEFAULT_ADVERTISEMENT = "[class,eq,'STOCK'],[symbol,isPresent,'A'],[open,isPresent,1],"
			+ "[high,isPresent,1],[low,isPresent,1],[close,isPresent,1],"
			+ "[volume,isPresent,1],[date,isPresent,'A']";

	private final static String DEFAULT_PUBLISHER_ID = "P" + DEFAULT_CLIENT_ID;

	private final static String DEFAULT_PUB_RATE = "1.0"; // msg/min

	private final static String DEFAULT_PUB_DATA_PATH = ClientConfig.PADRES_HOME
			+ "/demo/data/stockquotes/";

	private final static String DEFAULT_PUB_DELAY = "0";

	private final static long DELAY_BETWEEN_NEW_PUBLISHER = 1000; // ms

	// Command-line argument-related constants
	private static final String CMD_ARG_SYMBOL = "s";

	private static final String CMD_ARG_RATE = "r";

	private static final String CMD_ARG_PUB_DELAY = "d";

	// make sure .split() uses "\\" for special chars
	public static final String PUB_DELIMIT_CHAR = "/";

	private static final int AD_FACTOR = 1;

	// The stock that this publisher publishes
	private final String symbol;

	// Time in between publishing in milliseconds.
	private final long publishPeriod;

	private final double rate; // just for reference when logging

	// time to start publish after instantiation in ms
	private final long pubDelay;

	// Advertisement for this publisher. This doesn't change after it is initialized
	private final Advertisement advertisement;

	// Useful when we want to disconnect from a specified broker and when we want to send a
	// publication to all of our current connected brokers
	private final HashMap<String, String> brokerIdAdvIdMap;

	// Used for reading in publications to publish
	private BufferedReader reader;

	// The number of publication sent
	private long pubCount;

	// The number of times the entire publication file was published
	private int publicationRounds;

	private Publication lastPublished = null;

	// to read from a file or just to repeat one publication again and again
	// it is a hack for profiling experiments; keep it "false" for normal StockPublisher functions
	private boolean repeatPub = false;

	public StockPublisher(ClientConfig clientConfig, String pubSymbol, double pubRate, double delay)
			throws ClientException {
		super(clientConfig);
		symbol = pubSymbol;
		rate = pubRate;
		pubDelay = Math.round(delay * 1000.0); // s to ms
		publishPeriod = toMilliSecondPeriod(rate); // convert to pub/sec
		brokerIdAdvIdMap = new HashMap<String, String>(DEFAULT_CONNECTION_CAPACITY);
		pubCount = 0;
		publicationRounds = 0;
		initializeReader();
		// advertise to the brokers and update our bookkeeping
		advertisement = initializeAdvertisement();
		for (String brokerURI : getBrokerURIList()) {
			for(int i = 0; i < AD_FACTOR; i++){
				AdvertisementMessage advMsg = advertise(advertisement, brokerURI);
				brokerIdAdvIdMap.put(brokerURI, advMsg.getMessageID());
				try {
					Thread.sleep(DELAY_BETWEEN_NEW_PUBLISHER);
				} catch (InterruptedException e) {
					// TODO Auto-generated catch block
					e.printStackTrace();
				}
			}
		}
	}

	/**
	 * Time to publish!
	 */
	public void publish() {
		String publicationStr = null;
		if (repeatPub) {
			// this is just a hack for performance testing
			publicationStr = "[class,'STOCK'],[symbol,'STEM'],[open,3.26],[high,3.26],[low,3.15],[close,3.18],[volume,1630900],[date,'21-Dec-04']";
		} else {
			// If the end of the file is reached, go back to the first line
			try {
				if ((publicationStr = reader.readLine()) == null) {
					initializeReader();
					publicationRounds++;
					publicationStr = reader.readLine();
				}
			} catch (IOException e) {
				exceptionLogger.error(e.getMessage());
			}
		}
		publicationStr += ",[num," + pubCount + "]";

		// Send the publication to all brokers that we're connected to
		for (String brokerURI : brokerIdAdvIdMap.keySet()) {
			Publication pub = null;
			PublicationMessage pubId = null;
			
			try {
				pub = MessageFactory.createPublicationFromString(publicationStr);
				lastPublished = pub;
				pubId = null;
				pubId = super.publish(pub, brokerURI);
				exceptionLogger.debug("Sent publication (" + pubId + "): " + pub);
				pubCount++;
			} catch (ClientException e) {
				exceptionLogger.error(e.getMessage());
			} catch (ParseException e){
				exceptionLogger.error(e.getMessage());
			}
		}
	}

	/*
	 * (non-Javadoc)
	 * 
	 * @see universal.rmi.RMIUniversalClient#shutdown()
	 */
	public void shutdown() {
		StringBuffer output = new StringBuffer();
		output.append("Summary log:\n");
		output.append("Publisher ID  	: " + getClientID() + "\n");
		output.append("Publication   	: " + symbol + "\n");
		output.append("Publication Rate	: " + rate + "msgs/min\n");
		output.append("Broker History	: " + getBrokerHistory() + "\n");
		output.append("Messages sent 	: " + pubCount + "\n");
		output.append("Rounds        	: " + publicationRounds + "\n");
		output.append("Last published   : " + lastPublished + "\n");

		exceptionLogger.info(output);
	}

	/**
	 * Converts the publication rate from messages per minute to messages per second
	 * 
	 * @param msgsPerMinute
	 * @return
	 */
	private long toMilliSecondPeriod(double msgsPerMinute) {
		return Math.round(1000.0 / (msgsPerMinute / 60.0));
	}

	/*
	 * (non-Javadoc)
	 * 
	 * @see java.lang.Runnable#run()
	 */
	public void run() {
		pause(pubDelay);
		while (true) {
			// let's delay first so that we give the advertisement some time to propagate
			pause(publishPeriod);
			// Time to publish!
			publish();
		}
	}

	private void pause(long delay) {
		try {
			Thread.sleep(delay);
		} catch (Exception e) {
		}
	}

	/*
	 * Initializes the reader to start reading from the beginning of the file
	 */
	private void initializeReader() {
		reader = null;
		try {
			exceptionLogger.debug("Initialize reader count: " + publicationRounds);
			reader = new BufferedReader(new FileReader(DEFAULT_PUB_DATA_PATH + symbol));
		} catch (IOException e) {
			exceptionLogger.error(e.getMessage());
		}
	}

	/*
	 * Makes an advertisement by examining the first publication in the publication file
	 */
	private Advertisement initializeAdvertisement() {
		Advertisement adv = null;
		try {
			String pubString = reader.readLine();
			pubString += ",[num,0]";
			adv = Advertisement.toAdvertisement(pubString);
			// change op for attribute symbol from "isPresent" to "eq"
			((Predicate) adv.getPredicateMap().get("symbol")).setOp("eq");
		} catch (Exception e) {
			System.out.println("Failed: (Publisher) Error making advertisement.  Using default:"
					+ DEFAULT_ADVERTISEMENT);

			try {
				adv = MessageFactory.createAdvertisementFromString(DEFAULT_ADVERTISEMENT);
			} catch (ParseException e1) {
				exceptionLogger.error(e.getMessage());
				return null;
			}
		}

		return adv;
	}

	private static void showUsage() {
		System.out.printf("Usage: java StockPublisher -%s <id> -%s <broker_uri>\n\t"
				+ "-%s <symbol> -%s <pub_rate/min> -%s <pub_start_time>\n",
				ClientConfig.CLI_OPTION_ID, ClientConfig.CLI_OPTION_BROKER_LIST, CMD_ARG_SYMBOL,
				CMD_ARG_RATE, CMD_ARG_PUB_DELAY);
	}

	/**
	 * @param args
	 */
	public static void main(String[] args) {
		try {
			new LogSetup(null);
		} catch (LogException e) {
			e.printStackTrace();
			System.exit(1);
		}
		String[] cliKeys = ClientConfig.getCommandLineKeys();
		String[] fullCLIKeys = new String[cliKeys.length + 3];
		System.arraycopy(cliKeys, 0, fullCLIKeys, 0, cliKeys.length);
		fullCLIKeys[cliKeys.length] = CMD_ARG_SYMBOL + ":";
		fullCLIKeys[cliKeys.length + 1] = CMD_ARG_RATE + ":";
		fullCLIKeys[cliKeys.length + 2] = CMD_ARG_PUB_DELAY + ":";
		CommandLine cmdLine = new CommandLine(fullCLIKeys);
		try {
			cmdLine.processCommandLine(args);
		} catch (Exception e) {
			e.printStackTrace();
			System.exit(1);
		}
		String rateString = cmdLine.getOptionValue(CMD_ARG_RATE, DEFAULT_PUB_RATE);
		String stockString = cmdLine.getOptionValue(CMD_ARG_SYMBOL);
		String pubDelay = cmdLine.getOptionValue(CMD_ARG_PUB_DELAY, DEFAULT_PUB_DELAY);

		// Now that logging is set up, we can start logging
		Logger logger = Logger.getLogger(Client.class);
		// Exit if stock symbol is not specified
		if (stockString == null) {
			logger.fatal("Don't know what to publish because stock symbol not specified.");
			showUsage();
			System.exit(1);
			// Exit if no broker address is specified
		} else if (cmdLine.getOptionValue(ClientConfig.CLI_OPTION_BROKER_LIST) == null) {
			logger.fatal("Don't know which broker to connect to.");
			showUsage();
			System.exit(1);
		}

		String[] stockSymbols = stockString.split(PUB_DELIMIT_CHAR);
		String[] stockRates = rateString.split(PUB_DELIMIT_CHAR);

		if (stockSymbols.length != stockRates.length) {
			System.err.println("Number of stock symbols does not match with number of rates provided! "
					+ stockSymbols.length + " != " + stockRates.length);
			System.exit(1);
		}

		StockPublisher[] publishers = new StockPublisher[stockSymbols.length];
		Thread pubThread[] = new Thread[stockSymbols.length];
		for (int i = 0; i < stockSymbols.length; i++) {
			// Before logging and exiting with error, need to instantiate the publisher
			// class so that logging is initialized before we use it
			try {
				ClientConfig clientConfig = new ClientConfig();
				clientConfig.clientID = DEFAULT_PUBLISHER_ID;
				clientConfig.overwriteWithCmdLineArgs(cmdLine);
				clientConfig.clientID = clientConfig.clientID + "_" + i;
				//
				// System.out.println("clientConfig = " + clientConfig.toString());
				// System.out.println("stock = " + stockSymbols[i]);
				// System.out.println("rate = " + stockRates[i]);
				// System.out.println();

				publishers[i] = new StockPublisher(clientConfig, stockSymbols[i],
						Double.parseDouble(stockRates[i]), Double.parseDouble(pubDelay));
				// Let it publish forever until this program is killed
				pubThread[i] = new Thread(publishers[i]);
				pubThread[i].start();
				Thread.sleep(DELAY_BETWEEN_NEW_PUBLISHER);
			} catch (ClientException e) {
				System.out.println(e.getMessage());
				System.exit(1);
			} catch (InterruptedException e) {
			}
		}

		for (int i = 0; i < stockSymbols.length; i++) {
			try {
				pubThread[i].join();
			} catch (InterruptedException ie) {
				ie.printStackTrace();
			}
		}

		// Now wait forever
		final Object dummyObj = new Object();
		synchronized (dummyObj) {
			try {
				dummyObj.wait();
			} catch (InterruptedException ie) {
				ie.printStackTrace();
			}
		}
	}

}
